### ISSUE TITLE
Add a version of System.Windows.Forms.Application.ThreadException to System.Environment

### ISSUE THREAD

**Issue by AraHaan** [dotnet/corefx#35666]
I would like the ``ThreadException`` event that is in the ``System.Windows.Forms`` assembly at https://github.com/dotnet/winforms/ to be in ``System.Environment`` in the ``System.Runtime`` or ``System.Runtime.Extensions`` assemblies because of the fact that a console application that spauns new threads that can possible throw exceptions that are possible not handled are not handled forcing the user to reference winforms just for plugging into that event. Maybe having the event moved to the ``System.Environment`` class and have the ``Application`` class's version of the event be a stub for this one for backwards compat and not cut from the ``ThreadExceptionDialog`` there too. Just being forced to reference the Windows Forms dependency just for a event that might be executing from a faulting thread spawned from ``System.Threading.Tasks`` is a pain in the butt. I am filing this issue as it has a long time been a painful issue of mine. I would like this changed for others to be able to use too. Also thread exception handling is already a xplat part of windows forms anyway am I right? I think moving it and stubbing it where it is now should not break anything (or well should not).

I would take a crack at the moving here on this issue as well, I just do not know all the parts of the code in the winforms repo that it would take to relocate to the ``Environment`` class. I would like some help.

And not to mention having the ``ThreadException`` event only be in winforms seems like a flaw in the implementation of the framework because it could be deployed from any thread.

Sorry, I do not know who else to mention

Basically my idea is like so:

```cs
namespace System
{
    public static class Environment // or whatever the System.Environment class is.
    {
        // other code to the class.
        public static event EventHandler<ThreadExceptionEventArgs> ThreadException
        {
            add
            {
                // however this is implemented.
            }

            remove
            {
                // however this is implemented.
            }
        }
    }
}
```

or if it fits better in ``AppDomain`` or something. I just would like it relocated somewhere else that seems reasonable.
cc: @tannergooding

**Comment by danmosemsft** [comment]
@AraHaan can you give an example of code that throws an exception that you cannot conveniently handle without the `ThreadException` even handler?

**Comment by AraHaan** [comment]
It seems after testing that that handler is in fact only used after I use ``Application.Run`` with a form, and does not apply for console applications even if I explicitly add ``[STAThread]`` to the entrypoint. I assumed this incorrectly and that the cases actually are thrown in the normal ``UnhandledException`` handler in the default ``AppDomain``.

So, in this case for console and .NET Core < 3.0, and .NET standard I exclude Windows forms by adding a ``WITH_WINFORMS`` define to my package that is only defined in the .NET Framework and .NET Core 3 targets (and only support mini dumping when running on windows). I would like however some code to manually invoke dumping on other platforms and report back to them where the dump(s) and the log(s) with the stack trace are located at like how I do on windows in that package though.
